File: src/filesystem/FileSystem.ts
// ============================================================
// FluxionJS V3 — FileSystem Interface & Types
// ezEngine-inspired filesystem abstraction. Implementations
// sit behind this interface so the engine/editor never
// touches raw IPC or Node `fs` directly.
// ============================================================

// ── Path utilities ───────────────────────────────────────────

/** Normalize a path to forward slashes and collapse duplicates. */
export function normalizePath(p: string): string {
  return p.replace(/\\/g, '/').replace(/\/+/g, '/').replace(/\/$/, '');
}

/** Join path segments with forward-slash normalization. */
export function pathJoin(...parts: string[]): string {
  return normalizePath(parts.join('/'));
}

/** Get the directory portion of a path. */
export function pathDirname(p: string): string {
  const n = normalizePath(p);
  const idx = n.lastIndexOf('/');
  return idx >= 0 ? n.substring(0, idx) : '.';
}

/** Get the file name (with extension) from a path. */
export function pathBasename(p: string, stripExt?: string): string {
  const n = normalizePath(p);
  let base = n.substring(n.lastIndexOf('/') + 1);
  if (stripExt && base.endsWith(stripExt)) {
    base = base.substring(0, base.length - stripExt.length);
  }
  return base;
}

/** Extract the extension including the dot, e.g. ".png". Lower-cased. */
export function pathExtension(p: string): string {
  const base = pathBasename(p);
  const dot = base.lastIndexOf('.');
  return dot >= 0 ? base.substring(dot).toLowerCase() : '';
}

/**
 * Return true if `child` is inside `parent` directory (prevents path traversal).
 * Both paths are normalized before comparison.
 */
export function isInsidePath(child: string, parent: string): boolean {
  const nc = normalizePath(child).toLowerCase();
  const np = normalizePath(parent).toLowerCase();
  return nc === np || nc.startsWith(np + '/');
}

// ── File Info ────────────────────────────────────────────────

export interface FileInfo {
  /** File or directory name (no path). */
  name: string;
  /** Full absolute path (forward-slashed). */
  path: string;
  /** File size in bytes (0 for directories). */
  size: number;
  /** True if this entry is a directory. */
  isDirectory: boolean;
  /** Extension including dot, lower-cased. Empty for dirs. */
  extension: string;
  /** Last-modified timestamp (ms since epoch). */
  modifiedAt: number;
}

// ── Directory Entry (lightweight, returned by readDir / walkDir) ─────

export interface DirEntry {
  name: string;
  path: string;
  isDirectory: boolean;
  /** File size in bytes (populated by walkDir; 0 for directories). */
  size?: number;
  /** Last-modified timestamp in ms since epoch (populated by walkDir). */
  modifiedAt?: number;
}

// ── WalkDir options ──────────────────────────────────────────

export interface WalkDirOptions {
  /** Recurse into sub-directories. Default: true. */
  recursive?: boolean;
  /** Include hidden entries (names starting with '.'). Default: false. */
  includeHidden?: boolean;
  /** Maximum recursion depth (undefined = unlimited). */
  maxDepth?: number;
  /** Only return files whose extension is in this list (e.g. ['.png', '.ts']). */
  filterExtensions?: string[];
}

// ── File Watch Events ────────────────────────────────────────

export type FileWatchEventType = 'create' | 'change' | 'delete';

export interface FileWatchEvent {
  type: FileWatchEventType;
  path: string;
}

export type FileWatchCallback = (event: FileWatchEvent) => void;

// ── IFileSystem Interface ────────────────────────────────────
// Mirrors the Rust FileSystem trait so code written against this
// interface compiles and runs identically on Tauri, web (OPFS),
// and in-memory environments.

export interface IFileSystem {
  // ── Text I/O ──────────────────────────────────────────────
  readFile(path: string): Promise<string>;
  writeFile(path: string, data: string): Promise<void>;
  /** Append text to a file; creates the file if it does not exist. */
  appendFile(path: string, data: string): Promise<void>;
  /** Write atomically (write-to-temp then rename). Crash-safe. */
  writeFileAtomic(path: string, data: string): Promise<void>;

  // ── Binary I/O ────────────────────────────────────────────
  readBinary(path: string): Promise<ArrayBuffer>;
  writeBinary(path: string, data: ArrayBuffer): Promise<void>;
  /** Write binary atomically. */
  writeBinaryAtomic(path: string, data: ArrayBuffer): Promise<void>;

  // ── Directory ─────────────────────────────────────────────
  readDir(path: string): Promise<DirEntry[]>;
  /** Recursively enumerate all entries under path. */
  walkDir(path: string, opts?: WalkDirOptions): Promise<DirEntry[]>;
  mkdir(path: string): Promise<void>;

  // ── Queries ───────────────────────────────────────────────
  exists(path: string): Promise<boolean>;
  isFile(path: string): Promise<boolean>;
  isDir(path: string): Promise<boolean>;
  stat(path: string): Promise<FileInfo>;
  fileSize(path: string): Promise<number>;

  // ── Mutation ──────────────────────────────────────────────
  delete(path: string): Promise<void>;
  rename(oldPath: string, newPath: string): Promise<void>;
  copy(srcPath: string, destPath: string): Promise<void>;

  // ── Watching ──────────────────────────────────────────────
  watch(path: string, callback: FileWatchCallback): Promise<string>;
  unwatch(watchId: string): Promise<void>;

  // ── Dialogs (editor-only, may return null in headless/game) ──
  openFileDialog(filters?: FileDialogFilter[]): Promise<string | null>;
  saveFileDialog(filters?: FileDialogFilter[]): Promise<string | null>;
  openDirDialog(): Promise<string | null>;

  // ── Paths ─────────────────────────────────────────────────
  getAppDataPath(): Promise<string>;
  tempDir(): Promise<string>;

  // ── JSON helpers ──────────────────────────────────────────
  readJson<T = unknown>(path: string): Promise<T>;
  writeJson(path: string, value: unknown): Promise<void>;

  // ── Directory helper ──────────────────────────────────────
  ensureDir(path: string): Promise<void>;
}

export interface FileDialogFilter {
  name: string;
  extensions: string[];
}

// ── FileSystemBase ────────────────────────────────────────────
// Abstract base class that provides default implementations for
// all helper methods so concrete classes only override what they
// can do better (e.g. true atomic writes, native walk, etc.).

export abstract class FileSystemBase implements IFileSystem {
  // ── Abstract core — must be implemented ──────────────────
  abstract readFile(path: string): Promise<string>;
  abstract writeFile(path: string, data: string): Promise<void>;
  abstract readBinary(path: string): Promise<ArrayBuffer>;
  abstract writeBinary(path: string, data: ArrayBuffer): Promise<void>;
  abstract readDir(path: string): Promise<DirEntry[]>;
  abstract mkdir(path: string): Promise<void>;
  abstract exists(path: string): Promise<boolean>;
  abstract stat(path: string): Promise<FileInfo>;
  abstract delete(path: string): Promise<void>;
  abstract rename(oldPath: string, newPath: string): Promise<void>;
  abstract copy(srcPath: string, destPath: string): Promise<void>;
  abstract watch(path: string, callback: FileWatchCallback): Promise<string>;
  abstract unwatch(watchId: string): Promise<void>;
  abstract openFileDialog(filters?: FileDialogFilter[]): Promise<string | null>;
  abstract saveFileDialog(filters?: FileDialogFilter[]): Promise<string | null>;
  abstract openDirDialog(): Promise<string | null>;
  abstract getAppDataPath(): Promise<string>;

  // ── Default helper implementations ───────────────────────

  async appendFile(path: string, data: string): Promise<void> {
    const existing = (await this.exists(path)) ? await this.readFile(path) : '';
    await this.writeFile(path, existing + data);
  }

  async writeFileAtomic(path: string, data: string): Promise<void> {
    await this.writeFile(path, data);
  }

  async writeBinaryAtomic(path: string, data: ArrayBuffer): Promise<void> {
    await this.writeBinary(path, data);
  }

  async walkDir(path: string, opts?: WalkDirOptions): Promise<DirEntry[]> {
    const result: DirEntry[] = [];
    await this._walkRecursive(path, 0, opts ?? {}, result);
    return result;
  }

  private async _walkRecursive(
    path: string,
    depth: number,
    opts: WalkDirOptions,
    out: DirEntry[],
  ): Promise<void> {
    const entries = await this.readDir(path);
    for (const e of entries) {
      if (!opts.includeHidden && e.name.startsWith('.')) continue;
      if (!e.isDirectory && opts.filterExtensions?.length) {
        const ext = pathExtension(e.path);
        if (!opts.filterExtensions.includes(ext)) continue;
      }
      out.push(e);
      if (e.isDirectory && opts.recursive !== false) {
        if (opts.maxDepth === undefined || depth < opts.maxDepth) {
          await this._walkRecursive(e.path, depth + 1, opts, out);
        }
      }
    }
  }

  async isFile(path: string): Promise<boolean> {
    try { return !(await this.stat(path)).isDirectory; } catch { return false; }
  }

  async isDir(path: string): Promise<boolean> {
    try { return (await this.stat(path)).isDirectory; } catch { return false; }
  }

  async fileSize(path: string): Promise<number> {
    try { return (await this.stat(path)).size; } catch { return 0; }
  }

  async tempDir(): Promise<string> { return '/tmp'; }

  async readJson<T = unknown>(path: string): Promise<T> {
    return JSON.parse(await this.readFile(path)) as T;
  }

  async writeJson(path: string, value: unknown): Promise<void> {
    await this.writeFileAtomic(path, JSON.stringify(value, null, 2));
  }

  async ensureDir(path: string): Promise<void> {
    if (!(await this.exists(path))) await this.mkdir(path);
  }
}

File: src-tauri/src/file_system.rs
use tauri::command;
use std::path::Path;
use std::sync::Arc;
use sha2::Digest;
use base64::{Engine as _, engine::general_purpose::STANDARD};
use crate::utils::{DirEntry, FileStat};
use crate::fs::{NativeFs, FileSystem, CopyOptions};

// Helper: convert a FileEntry list to the legacy DirEntry format used by
// the frontend (camelCase, flat path string).
fn to_dir_entries(entries: Vec<crate::fs::FileEntry>) -> Vec<DirEntry> {
    entries
        .into_iter()
        .map(|e| DirEntry {
            name:         e.name,
            is_directory: e.kind.is_directory(),
            path:         e.path.to_string_lossy().into_owned(),
            size:         e.size,
            modified_at:  e.modified_at,
        })
        .collect()
}

// ── Text I/O ─────────────────────────────────────────────────────────────────

#[command]
pub async fn read_file(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<String, String> {
    state.read_text(Path::new(&path)).map_err(|e| e.to_string())
}

#[command]
pub async fn write_file(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
    data:  String,
) -> Result<(), String> {
    state.write_text(Path::new(&path), &data).map_err(|e| e.to_string())
}

// ── Binary I/O (Base64 transport) ────────────────────────────────────────────

#[command]
pub async fn read_binary(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<String, String> {
    let bytes = state.read_bytes(Path::new(&path)).map_err(|e| e.to_string())?;
    Ok(STANDARD.encode(&bytes))
}

#[command]
pub async fn write_binary(
    state:       tauri::State<'_, Arc<NativeFs>>,
    path:        String,
    base64_data: String,
) -> Result<(), String> {
    let data = STANDARD.decode(&base64_data)
        .map_err(|e| format!("Failed to decode base64: {e}"))?;
    state.write_bytes(Path::new(&path), &data).map_err(|e| e.to_string())
}

// ── Directory ─────────────────────────────────────────────────────────────────

#[command]
pub async fn list_dir(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<Vec<DirEntry>, String> {
    let entries = state.list_dir(Path::new(&path)).map_err(|e| e.to_string())?;
    Ok(to_dir_entries(entries))
}

#[command]
pub async fn read_dir(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<Vec<DirEntry>, String> {
    list_dir(state, path).await
}

#[command]
pub async fn mkdir(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<(), String> {
    state.mkdir(Path::new(&path)).map_err(|e| e.to_string())
}

// ── Queries ───────────────────────────────────────────────────────────────────

#[command]
pub async fn exists(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<bool, String> {
    Ok(state.exists(Path::new(&path)))
}

#[command]
pub async fn stat(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<FileStat, String> {
    let s = state.stat(Path::new(&path)).map_err(|e| e.to_string())?;
    Ok(FileStat {
        size:         s.size,
        is_directory: s.kind.is_directory(),
        modified_at:  s.modified_at,
    })
}

#[command]
pub async fn is_file(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<bool, String> {
    Ok(state.is_file(Path::new(&path)))
}

#[command]
pub async fn is_directory(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<bool, String> {
    Ok(state.is_dir(Path::new(&path)))
}

// ── Mutation ──────────────────────────────────────────────────────────────────

#[command]
pub async fn delete_file(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<(), String> {
    state.delete(Path::new(&path)).map_err(|e| e.to_string())
}

#[command]
pub async fn rename(
    state:    tauri::State<'_, Arc<NativeFs>>,
    old_path: String,
    new_path: String,
) -> Result<(), String> {
    state.rename(Path::new(&old_path), Path::new(&new_path))
        .map_err(|e| e.to_string())
}

#[command]
pub async fn copy(
    state:     tauri::State<'_, Arc<NativeFs>>,
    src_path:  String,
    dest_path: String,
) -> Result<(), String> {
    state.copy_file(
        Path::new(&src_path),
        Path::new(&dest_path),
        &CopyOptions { overwrite: true, skip_existing: false },
    ).map_err(|e| e.to_string())
}

// ── Hashing ───────────────────────────────────────────────────────────────────

#[command]
pub async fn hash_file(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<String, String> {
    let data = state.read_bytes(Path::new(&path)).map_err(|e| e.to_string())?;
    let mut hasher = sha2::Sha256::new();
    hasher.update(&data);
    Ok(format!("{:x}", hasher.finalize()))
}

// ── Convenience ───────────────────────────────────────────────────────────────

#[command]
pub async fn get_file_size(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
) -> Result<u64, String> {
    Ok(state.file_size(Path::new(&path)))
}

// ── Atomic writes ─────────────────────────────────────────────────────────────

#[command]
pub async fn append_file(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
    data:  String,
) -> Result<(), String> {
    state.append_text(Path::new(&path), &data).map_err(|e| e.to_string())
}

#[command]
pub async fn write_file_atomic(
    state: tauri::State<'_, Arc<NativeFs>>,
    path:  String,
    data:  String,
) -> Result<(), String> {
    state.write_text_atomic(Path::new(&path), &data).map_err(|e| e.to_string())
}

#[command]
pub async fn write_binary_atomic(
    state:       tauri::State<'_, Arc<NativeFs>>,
    path:        String,
    base64_data: String,
) -> Result<(), String> {
    let data = STANDARD.decode(&base64_data)
        .map_err(|e| format!("Failed to decode base64: {e}"))?;
    state.write_bytes_atomic(Path::new(&path), &data).map_err(|e| e.to_string())
}

// ── Walk directory ────────────────────────────────────────────────────────────

#[command]
pub async fn walk_dir_cmd(
    state:          tauri::State<'_, Arc<NativeFs>>,
    path:           String,
    recursive:      Option<bool>,
    include_hidden: Option<bool>,
    max_depth:      Option<usize>,
    filter_exts:    Option<Vec<String>>,
) -> Result<Vec<DirEntry>, String> {
    use crate::fs::WalkOptions;
    let opts = WalkOptions {
        include_hidden:   include_hidden.unwrap_or(false),
        max_depth:        if recursive.unwrap_or(true) { max_depth } else { Some(0) },
        filter_extensions: filter_exts.unwrap_or_default(),
    };
    let entries = state.walk_dir(Path::new(&path), &opts).map_err(|e| e.to_string())?;
    Ok(entries.into_iter().map(|e| DirEntry {
        name:         e.name,
        is_directory: e.kind.is_directory(),
        path:         e.path.to_string_lossy().into_owned(),
        size:         e.size,
        modified_at:  e.modified_at,
    }).collect())
}

// ── Platform paths ────────────────────────────────────────────────────────────

#[command]
pub async fn get_temp_dir(
    state: tauri::State<'_, Arc<NativeFs>>,
) -> Result<String, String> {
    Ok(state.temp_dir().to_string_lossy().into_owned())
}
